# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/frontend/ForOfLoopControl.cpp
# Commit: eeaeb40f8b86
# Full Hash: eeaeb40f8b869121eff65b1c2d11af423c9a9ecc
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2022-03-09 09:44:44
# Regressor Bug: 885514
# File Overlap Count: 1
# Description:
#   Bug 885514: Simplify JSOp::Retsub r=jandem
#   
#   Currently, JSOp::Retsub takes two arguments, and uses one to decide whether to throw the second or use it as a resume index. This patch splits out the exception-handling logic into explicit bytecode, leaving JSOp::Retsub as a simple indirect jump.
#   
#   To make the generated bytecode work out nicely, I've swapped the order of the arguments: now `throwing` is on top, and the exception/resumeindex argument is underneath it.
# ==============================================================================

diff -r 1cb0b352b024 -r eeaeb40f8b86 js/src/frontend/ForOfLoopControl.cpp
--- a/js/src/frontend/ForOfLoopControl.cpp	Tue Mar 08 21:10:22 2022 +0000
+++ b/js/src/frontend/ForOfLoopControl.cpp	Tue Mar 08 21:10:22 2022 +0000
@@ -89,14 +89,21 @@
 
   // If any yields were emitted, then this for-of loop is inside a star
   // generator and must handle the case of Generator.return. Like in
-  // yield*, it is handled with a finally block.
+  // yield*, it is handled with a finally block. If the generator is
+  // closing, then the exception/resumeindex value (second value on
+  // the stack) will be a magic JS_GENERATOR_CLOSING value.
+  // TODO: Refactor this to eliminate the swaps.
   uint32_t numYieldsEmitted = bce->bytecodeSection().numYields();
   if (numYieldsEmitted > numYieldsAtBeginCodeNeedingIterClose_) {
     if (!tryCatch_->emitFinally()) {
       return false;
     }
-
+    //              [stack] ITER ... FVALUE FTYPE
     InternalIfEmitter ifGeneratorClosing(bce);
+    if (!bce->emit1(JSOp::Swap)) {
+      //            [stack] ITER ... FTYPE FVALUE
+      return false;
+    }
     if (!bce->emit1(JSOp::IsGenClosing)) {
       //            [stack] ITER ... FTYPE FVALUE CLOSING
       return false;
@@ -118,6 +125,10 @@
       //            [stack] ITER ... FTYPE FVALUE
       return false;
     }
+    if (!bce->emit1(JSOp::Swap)) {
+      //            [stack] ITER ... FVALUE FTYPE
+      return false;
+    }
   }
 
   if (!tryCatch_->emitEnd()) {